### DISCUSSION TITLE
One Time Password hook

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#5482] [Show and tell]
Here is an open source hook to help with one-time passwords https://www.npmjs.com/package/pocketbase-otp

PR's welcome!

## Installation

1. Install the package:

   ```bash
   npm install pocketbase-otp
   ```

2. Copy the migrations into your `pb_migrations`:

   ```bash
   cp node_modules/pocketbase-otp/pb_migrations pb_migrations
   ```

3. Copy the hook file into your `pb_hooks`::

   ```bash
   cp node_modules/pocketbase-otp/pb_hooks pb_hooks
   ```

### Requesting a Code

First, make sure your PocketBase instance has SMTP configured and tested for sending emails.

To request an OTP code:

```javascript
const client = new PocketBase(...)
const res = await client.send(`/api/otp/auth`, {
  body: { email },
  method: 'POST',
})
```

The PocketBase instance will send the OTP code to the provided email address.

### Verifying the Code

Once the user receives the OTP code, use it to trade for the authentication token and user record:

```javascript
const res = await client.send(`/api/otp/verify`, {
  body: { email, code },
  method: 'POST',
})
client.authStore.save(res.token, res.record)
```

This will authenticate the PocketBase client using the OTP.

**Comment by ganigeorgiev** [comment]
_Side-note: OTP (and MFA) support is already implemented as part of the ongoing refactoring and will be available with the next v0.23.0 soon. Additionally I'm not sure if such things makes sense to be distributed as npm packages, or at least not in its current form, because based on your installation instructions users have to run manual `cp` commands, which kind of defeat the purpose of using npm. Instead a better approach maybe would be to export your code as regular CJS module and add instruction for users to just call something like `const myhookInit = require("myhook"); myhookInit()` where `myHookInit` is a function that binds to the necessary hooks (the migrations in this case could be applied as part of the after bootstrap hook with `if (!collectionExists) { create... }` check)._

**Reply by benallfree** [reply]
> where myHookInit is a function that binds to the necessary hooks (the migrations in this case could be applied as part of the after bootstrap hook with if (!collectionExists) { create... } check).

Good idea!

> will be available with the next v0.23.0 soon

👀

**Reply by benallfree** [reply]
Gani I've been thinking more and this could still be a valuable conversation for an emerging plugin ecosystem.

Could I hear more about how to manage migrations? For example, a plugin would need to be able to uninstall itself too. Could I use the settings table to track migration version numbers so I could potentially upgrade and uninstall? I am imagining that the normal approach of using migration files would not be suitable for something that might need to be uninstalled/reinstalled.

**Reply by ganigeorgiev** [reply]
Migrations are usually intended for userland code. I'm not sure if adding external code as part of the app migrations is a good idea as it could cause accidental data loss on `migrate down` because there is no guarantee that the plugin migration will be the last one to revert.

> a plugin would need to be able to uninstall itself too

It seems that you want a "WordPress" style of extensibility but Go is compiled language and loading/unloading code dynamically during runtime is not an easy task (even with goja). You are free to handle it however you think suit your needs but keep in mind that PocketBase is not stable yet and there will be breaking changes. Some sort of official Go/JSVM plugin guidelines will be available eventually in the future but I don't have time to think about this at the moment, nor I consider it a priority.
